### PR TITLE
Feature parametric types

### PR DESCRIPTION
**Pull request by pnowojski**
This is part of the work for the following issue: https://github.com/facebook/presto/issues/3938
- generailized parametric types
- deprecated literalParameters
- added parser support for BIGINT typeParameter
- propagated TypeParameterSignature to most of previous TypeSignature uses

TODO/Questions:
- check/add backward compatibility with old clients (changes in ClientTypeSignature)
- move parseSignature to presto-main and base it on antlr

### CODE REVIEW

**Comment by pnowojski** [comment]
I am not entirely sure if that's right direction. It seems like having two separate lists for literal parameters and type parameters was simpler from code's perspective. Do we really want to unify both of those things? Without unifying we are losing the way to distinguish between two types `foo(BIGINT, 42)` and `foo(42, BIGINT)` but do we need that?

**Comment by losipiuk** [comment]
some comments

**Comment by pnowojski** [comment]
@losipiuk yes - If I didn't make your suggested change I made some comment, so if there is no comment, it was fixed.

( yes, as you know I'm also a big fan of the github as a code reviewing platform... )

**Comment by pnowojski** [comment]
Squashed all commits, because there was some overlapping between them and this pull request is not that big

**Comment by pnowojski** [comment]
@martint I have changed implementation of `*Parameter` classes to your proposed one and I have pushed two more commits that moved ROW type to generalized type parameters (thus completely dropping old `literalParameters`).

It would be best to minimize changes to commits from this pull request, since there is a lot of more code that depends on those three commits (rebased Dain's VARCHAR changes + decimal changes).

We still need to decide at least what to do with backward compatibility with previous clients.

As a future refactor somewhere down the road, we should indeed do something about redundancy between TypeSignatureParameter, TypeParameter and ClientTypeSignatureParameter classes. Do we need ClientTypeSignature at all? It predates dependency of presto-client to presto-spi, so it might HAD some sense long time ago, but at this moment, I don't see a point of having TypeSignature and ClientTypeSignature?

**Comment by martint** [comment]
Great, I'll take a look.

I don't think we need ClientTypeSignature any more, by the way.

File: presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4 [hunk 1]
```diff
@@ -238,7 +238,7 @@ valueExpression
 primaryExpression
     : NULL                                                                           #nullLiteral
     | interval                                                                       #intervalLiteral
-    | identifier STRING                                                              #typeConstructor
+    | type STRING                                                                    #typeConstructor
```
Comment (martint) [inline presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4]: I'm not sure we want to allow arbitrary types for the "type constructor" idiom. I actually think it was a mistake to allow anything beyond the special cases described in the SQL spec (INTERVAL, etc).

Comment (pnowojski) [inline presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4]: Isn't this out of scope of this pull request? Can we create an issue for "disallow arbitrary types in type constructor"? Especially that @losipiuk said that such solution was discussed with you about constructing decimal literals.

**Comment by martint** [comment]
> Isn't this out of scope of this pull request? Can we create an issue for "disallow arbitrary types in type constructor"

What I was suggesting is not that we change what we support today (at least not yet), but to not extend it to arbitrary types. These look weird to me: `VARCHAR(10) 'hello'`, `DECIMAL(5,2) '123.456'`.

Is there a use case for specifying the type parameters in these expressions given that it can be inferred from the value?

**Comment by martint** [comment]
Superseded by #4264

**Pull request closed without merging**